### PR TITLE
Updated to use proc_macro (fixes #8).

### LINKED ISSUE #8: Switch to proc_macro
https://github.com/rust-lang/rust/issues/38356

It's going to be stable soon. Finally we can stop chasing changes in the private rustc crates' APIs. Probably a good time to bump to v1.0.0 at the same time.

### PR DESCRIPTION
**Pull request by 6A**
The following breaking changes are therefore introduced:
- Switch from plugin to proc_macro, which means that `speculate` must be imported as you'd import a function (`use speculate::speculate`), and no longer as a plugin (`#![plugin(speculate)]`).
- Identifiers can no longer contain special characters or spaces; as thus they're filtered, so that the test named `0 is... illegal!!` is compiled as `_0_is_illegal`.
- The generated module can no longer be called `speculate`, as it conflicts with the name of the crate. Instead, the created modules are named `speculate_tests`.
- The `speculate` macro can once again only be used once in a scope. The macro `speculate_again` was added to allow multiple uses in a same scope. With the latter, the `_line_[line]` text is added to the module name.

Internally, `syn`, `quote` and `proc_macro2` are now used for parsing / generating Rust code.

Also, I think the version should be incremented in `Cargo.toml` (to at least 0.1.0, since this is a breaking change). Maybe I should also add my name to the authors field?

**Note**: The README hasn't been updated for the proc_macro usage yet... Should I do it?

### CODE REVIEW

**Comment by utkarshkukreti** [comment]
Thanks a lot for this! I'll try to review and merge as soon as I can.

> Also, I think the version should be incremented in Cargo.toml (to at least 0.1.0, since this is a breaking change). Maybe I should also add my name to the authors field?

Yes, please bump to 0.1.0 and add yourself to the authors field :)

**Comment by utkarshkukreti** [comment]
What do you think about removing `speculate_again` and changing `speculate` to generate a module named `"speculate_" + line_number`? This will be almost what we have right now (we do `"speculate #" + byte_offset` right now) plus no need of two macros.

**Comment by 6A** [comment]
Alright, I'm going back to `speculate_<line number>` by default! I'm also changing the README, and adding docs to the module / macro.

**Comment by 6A** [comment]
Oh, and I noticed you're mentioning the MIT license, but there is no license file in the repository. I think you should include one.

File: README.md [hunk 1]
```diff
@@ -9,18 +9,22 @@ between stable and nightly versions of Rust.
 
 ## Installation
 
-Add `speculate` to the `dev-dependencies` section of your `Cargo.toml`:
+Add `speculate` to the `dependencies` section of your `Cargo.toml`:
```
Comment (utkarshkukreti) [inline README.md]: Is this change required?

**Comment by utkarshkukreti** [comment]
I tried updating https://github.com/utkarshkukreti/select.rs to this but I'm getting some strange errors:

```
error[E0061]: this function takes 0 parameters but 1 parameter was supplied
   --> tests/predicate_tests.rs:12:1
    |
12  |   speculate! {
    |   ^---------
    |   |
    |  _defined here
    | |
13  | |     describe "predicate" {
14  | |         before {
15  | |             let document = Document::from("<html><head></head><body>\
...   |
179 | |     }
180 | | }
    | |_^ expected 0 parameters
```

Do you have any idea about this? The code is in this branch: https://github.com/utkarshkukreti/select.rs/tree/speculate-0.1.0. You'll need to change the path in `Cargo.toml` to point to a local checkout of this PR.

File: README.md (hunk 1 above)
Comment (6A) [reply README.md]: Not really no. However with `proc-macro`, the boilerplate is a tad more annoying:

```rust
// lib.rs:
#![cfg_attr(test, feature(plugin))]
#![cfg_attr(test, plugin(speculate))]

// any module:
#[cfg(test)]
speculate! {
    ...
}
```

becomes....

```rust
// lib.rs:
#[cfg(test)]
extern crate speculate;

// any module:
#[cfg(test)]
use speculate::speculate;

#[cfg(test)]
speculate! {
    ...
}
```

But I suppose it's not a big enough change to recommend `dependencies` instead of `dev-dependencies`.

**Comment by 6A** [comment]
Okay, I know what's happening in `select.rs`.

## Span informations are not preserved
Stable `proc-macro` does not have span informations, therefore when an error is reported, it is reported for the whole macro block. Adding `proc-macro2 = { version = "=0.4.8", features = [ "nightly" ] }` to `Cargo.toml` makes error reporting more accurate.

Example:

```
error[E0061]: this function takes 0 parameters but 1 parameter was supplied
  --> tests\predicate_tests.rs:64:24
   |
60 |         test "Not()" {
   |              ------- defined here
...
64 |             assert_eq!(Not(Name("head")).matches(&head), false);
   |                        ^^^^^^^^^^^^^^^^^ expected 0 parameters
```

## Function names such as `Name` and `Attr` conflict with the generated module names
Since `proc-macro` does not allow special characters in identifiers, the test `test "Not()"` will generate a function named `Not`.

Example

```rust
test "Not()" {
    assert_eq!(Not(Name("html")).matches(&html), false);
    assert_eq!(Not(Name("html")).matches(&head), true);
    assert_eq!(Not(Name("head")).matches(&html), true);
    assert_eq!(Not(Name("head")).matches(&head), false);
}

// becomes

#[test]
fn Not() {
    // Here 'Not' refers to the test function, instead of the struct.
    assert_eq!(Not(Name("html")).matches(&html), false);
    assert_eq!(Not(Name("html")).matches(&head), true);
    assert_eq!(Not(Name("head")).matches(&html), true);
    assert_eq!(Not(Name("head")).matches(&head), false);
}
```

I guess we can fix this by appending `_test` or something similar to the generated function name.

**Comment by 6A** [comment]
I pushed my changes, but adding the `nightly` feature to `proc-macro2` means speculate will only work on the very latest versions of the nightly compiler...

**Comment by utkarshkukreti** [comment]
> I pushed my changes, but adding the nightly feature to proc-macro2 means speculate will only work on the very latest versions of the nightly compiler...

The crate already only works on latest nightlies so that's fine. I'll create an issue for doing a 1.0.0 release once all features we use are stable.

> Also, the output is slightly more verbose:

Yeah, I think that's fine. I was originally going to go with `test_$message` but not sure why I went with `$message`. The reason the test for `Not` in select.rs is called `Not()` and not `Not` is to prevent name clash. I can switch that to `Not` now.


Everything looks good, thanks, merging!
